File: frontend/src/lib/api/teacher.ts
import { apiClient } from "./client";
import { TeacherSummaryResponse, StudentDetailResponse } from "@/types/api";

/**
 * Frontend function: getTeacherInsights
 * HTTP method: GET
 * Backend URL: /api/v1/teacher/insights
 * Request type: Query params (grade, subject, concept_id, status)
 * Response type: TeacherSummaryResponse
 */
export async function getTeacherInsights(params?: {
  grade?: number;
  subject?: string;
  concept_id?: string;
  status?: string;
}): Promise<TeacherSummaryResponse> {
  const urlParams = new URLSearchParams();
  if (params) {
    if (params.grade !== undefined) urlParams.append("grade", params.grade.toString());
    if (params.subject) urlParams.append("subject", params.subject);
    if (params.concept_id) urlParams.append("concept_id", params.concept_id);
    if (params.status) urlParams.append("status", params.status);
  }

  const queryString = urlParams.toString();
  const endpoint = `/api/v1/teacher/insights${queryString ? `?${queryString}` : ""}`;

  return apiClient<TeacherSummaryResponse>(endpoint, {
    method: "GET",
  });
}

/**
 * Frontend function: getStudentInsights
 * HTTP method: GET
 * Backend URL: /api/v1/teacher/students/{student_id}/insights
 * Request type: Path param (student_id)
 * Response type: StudentDetailResponse
 */
export async function getStudentInsights(
  studentId: string
): Promise<StudentDetailResponse> {
  return apiClient<StudentDetailResponse>(
    `/api/v1/teacher/students/${encodeURIComponent(studentId)}/insights`,
    {
      method: "GET",
    }
  );
}

File: backend/app/api/v1/teacher.py
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from app.db.database import get_db

from app.api.schemas.teacher import TeacherSummaryResponse, StudentDetailResponse
from app.services.teacher_insights import TeacherInsightService
from app.api.deps import get_current_teacher, CurrentUser

router = APIRouter()

def get_insight_service():
    return TeacherInsightService()

@router.get("/insights", response_model=TeacherSummaryResponse)
def get_teacher_insights(
    grade: Optional[int] = Query(None, description="Filter by student grade"),
    subject: Optional[str] = Query(None, description="Filter by concept subject"),
    concept_id: Optional[str] = Query(None, description="Filter by specific concept ID"),
    status: Optional[str] = Query(None, description="Filter by insight status (e.g. at_risk)"),
    db: Session = Depends(get_db),
    service: TeacherInsightService = Depends(get_insight_service),
    current_teacher: CurrentUser = Depends(get_current_teacher)
):
    try:
        return service.get_class_summary(
            db=db,
            grade=grade,
            subject=subject,
            concept_id=concept_id,
            status_filter=status
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/students/{student_id}/insights", response_model=StudentDetailResponse)
def get_student_insights(
    student_id: str,
    db: Session = Depends(get_db),
    service: TeacherInsightService = Depends(get_insight_service),
    current_teacher: CurrentUser = Depends(get_current_teacher)
):
    try:
        return service.get_student_detail(db=db, student_id=student_id)
    except ValueError as ve:
        raise HTTPException(status_code=404, detail=str(ve))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
